File: algorithms/M/Maximum Number of Removable Characters/Maximum Number of Removable Characters.py
class Solution:
    def maximumRemovals(self, s: str, p: str, removable: List[int]) -> int:
        
        def check(m):
            i = j = 0
            remove = set(removable[:m+1])
            while i < len(s) and j < len(p):
                if i in remove:
                    i += 1
                    continue
                if s[i] == p[j]:
                    i += 1
                    j += 1
                else:
                    i += 1
            
            return j == len(p)
            
                
        # search interval is [lo, hi)
        lo, hi = 0, len(removable)+1
        
        while lo < hi:
            mid = (lo + hi) // 2
            if check(mid):
                lo = mid + 1
            else:
                hi = mid
                
        return lo if lo < len(removable) else lo-1

File: algorithms/M/Maximum Number of Removable Characters/Maximum Number of Removable Characters.js
var maximumRemovals = function(s, p, removable) {
    let arr = s.split('');
    
    const stillFunctions = (k) => {
        let result = [...arr];
        for(let i = 0; i < k; i++) {
            result[removable[i]] = '';
        }

        const isSubset = () => {
            let idx = 0;
            for(let c = 0; c < p.length; c++) {
                if(idx > result.length) return false;
                const next = result.indexOf(p[c], idx);
                if(next === -1) {
                    return false;
                }
                idx = next + 1;
            }
            return true;
        }
        
        return isSubset();
    }

    let left = 0;
    let right = removable.length;
    
    while(left < right) {
        // Need to round up the midpoint since we are returning operation LENGTH i.e. ops+1
        const mid = Math.ceil((left+right) / 2);
        if(!stillFunctions(mid)) {
            right = mid -1;
        } else {
            left = mid;
        }
    }
    return left;
};
